dds/DCPS/XTypes/TypeObject.cpp

```diff
@@ -471,616 +471,719 @@ const TypeMap TypeMapBuilder::EmptyMap;
-    break;
-  case TI_STRING8_LARGE:
-  case TI_STRING16_LARGE:
-    compute_dependencies_i(type_map, type_identifier.string_ldefn(), dependencies);
-    break;
-  case TI_PLAIN_SEQUENCE_SMALL:
-    compute_dependencies_i(type_map, type_identifier.seq_sdefn(), dependencies);
-    break;
-  case TI_PLAIN_SEQUENCE_LARGE:
-    compute_dependencies_i(type_map, type_identifier.seq_ldefn(), dependencies);
-    break;
-  case TI_PLAIN_ARRAY_SMALL:
-    compute_dependencies_i(type_map, type_identifier.array_sdefn(), dependencies);
-    break;
-  case TI_PLAIN_ARRAY_LARGE:
-    compute_dependencies_i(type_map, type_identifier.array_ldefn(), dependencies);
-    break;
-  case TI_PLAIN_MAP_SMALL:
-    compute_dependencies_i(type_map, type_identifier.map_sdefn(), dependencies);
-    break;
-  case TI_PLAIN_MAP_LARGE:
-    compute_dependencies_i(type_map, type_identifier.map_ldefn(), dependencies);
-    break;
-  case TI_STRONGLY_CONNECTED_COMPONENT:
-    compute_dependencies_i(type_map, type_identifier.sc_component_id(), dependencies);
-    break;
-  case EK_COMPLETE:
-  case EK_MINIMAL:
-    {
-      TypeMap::const_iterator pos = type_map.find(type_identifier);
-      if (pos != type_map.end()) {
-        compute_dependencies_i(type_map, pos->second, dependencies);
+  void compute_dependencies_j(const TypeMap& type_map,
+                              const TypeIdentifier& type_identifier,
+                              TypeIdentifierSet& dependencies,
+                              TypeIdentifierSet& sccs)
+  {
+    switch (type_identifier.kind()) {
+    case TI_STRING8_SMALL:
+    case TI_STRING16_SMALL:
+      compute_dependencies_i(type_map, type_identifier.string_sdefn(), dependencies, sccs);
+      break;
+    case TI_STRING8_LARGE:
+    case TI_STRING16_LARGE:
+      compute_dependencies_i(type_map, type_identifier.string_ldefn(), dependencies, sccs);
+      break;
+    case TI_PLAIN_SEQUENCE_SMALL:
+      compute_dependencies_i(type_map, type_identifier.seq_sdefn(), dependencies, sccs);
+      break;
+    case TI_PLAIN_SEQUENCE_LARGE:
+      compute_dependencies_i(type_map, type_identifier.seq_ldefn(), dependencies, sccs);
+      break;
+    case TI_PLAIN_ARRAY_SMALL:
+      compute_dependencies_i(type_map, type_identifier.array_sdefn(), dependencies, sccs);
+      break;
+    case TI_PLAIN_ARRAY_LARGE:
+      compute_dependencies_i(type_map, type_identifier.array_ldefn(), dependencies, sccs);
+      break;
+    case TI_PLAIN_MAP_SMALL:
+      compute_dependencies_i(type_map, type_identifier.map_sdefn(), dependencies, sccs);
+      break;
+    case TI_PLAIN_MAP_LARGE:
+      compute_dependencies_i(type_map, type_identifier.map_ldefn(), dependencies, sccs);
+      break;
+    case TI_STRONGLY_CONNECTED_COMPONENT:
+      {
+        // An SCCId (identifies the entire component.)
+        TypeIdentifier scc_id = make_scc_id(type_identifier);
+        sccs.insert(scc_id);
```

Comment:
The returned `scc_id` has `scc_index` value of 0. I think the SCC index starts from 1?